Make hashes, length and delegations optional + improvements

MVrachev opened the pull request:
Fixes: #1362

**Description of the changes being introduced by the pull request**:

**Please note: this pr contains one breaking change:**
1. make `delegations` optional for `Targets`

The idea of this pr came from improvements proposed/discussed in pr #1329.
Those improvements are proposed together (because they are small enough to be bundled) in a
separate pr and with that to simplify the review on pr #1329.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
New API: Add root use case in couple of tests

Add a use case for the root class to be tested in test_generic_read
and test_read_write_read_compare tests in test_apy.py

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
NEW API: Fix documentation indentation

From the reST/sphinx docs:
https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks

I added new lines and an identation where it was missed.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Make length and hashes optional in Timestamp

As per the specification (v1.0.1) length and hashes fields
in timestamp and snapshot metadata are optional.
We have implement this in the older API
(see https://github.com/theupdateframework/tuf/pull/1031) and we should
implement it in the new API.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
BREAKING CHANGE: Make delegations optional

According to the spec, delegations in targets are marked as optional:
https://theupdateframework.github.io/specification/latest/#file-formats-targets
and a pr, clarifying that even more, is approved:
https://github.com/theupdateframework/specification/pull/157.

This is a possible breaking change.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

jku commented on tuf/api/metadata.py:
```diff
@@ -336,6 +336,8 @@ def __init__(
         unrecognized_fields: Optional[Mapping[str, Any]] = None,
     ) -> None:
 
+        # "_type" is not actually meant to be "protected" in the OOP sense
+        # but rather is a workaround to not shadow the built-in "type" name.
```
I've been wondering, is this actually true for the object attribute?  As in, I get that arguments/variables shouldn't be named `type` but why can't the attribute be `Signed.type` -- it doesn't shadow anything?

I don't think `Signed.type` would be confusing to user (in fact I'd argue it's in line with imitating the file format) and I don't think there's actually anything preventing us from calling it that as `type` is not a reserved keyword... Maybe I'm missing a tool that complains about this?

MVrachev replied:
I searched for that and one relevant response from Stackoverflow: https://stackoverflow.com/a/10568115/7058471.
I am not against or for renaming.

joshuagl replied:
One other reason to use `_type` rather than `type` is that it better matches the spec. 

Historical context: the use of `_type` in the spec was to match the implementation, which did not want to shadow the builtin `type()` method. Shadowing `type()` was more problematic in a procedural implementation, and will likely be less of an issue in the OOP metadata API – but the name in the spec is `_type` until we do something with https://github.com/theupdateframework/specification/issues/149 and a possible 2.0 of the specification.

jku replied:
oof I forgot the spec uses _type . Thanks

jku commented on tests/test_api.py:
```diff
@@ -258,6 +258,10 @@ def test_metadata_snapshot(self):
+        # Update only version. Length and hashes are optional.
+        snapshot.signed.update('role1', 3)
+        fileinfo['role1.json'] = {'version': 3}
+        self.assertEqual(snapshot.signed.meta, fileinfo)
```
A de/serialize test with json without length/hashes would be good for future

jku commented on tests/test_api.py:
```diff
@@ -293,6 +297,10 @@ def test_metadata_timestamp(self):
         timestamp.signed.update(2, 520, hashes)
         self.assertEqual(timestamp.signed.meta['snapshot.json'], fileinfo)
 
+        # Update only version. Length and hashes are optional.
+        timestamp.signed.update(3)
+        fileinfo = {'version': 3}
+        self.assertEqual(timestamp.signed.meta['snapshot.json'], fileinfo)
```
A de/serialize test with json without length/hashes would be good for future

jku commented on tuf/api/metadata.py:
```diff
@@ -557,14 +557,19 @@ def to_dict(self) -> Dict[str, Any]:
-            "length": length,
-            "hashes": hashes,
-        }
+        self.meta["snapshot.json"] = {"version": version}
+
+        if length is not None:
+            self.meta["snapshot.json"]["length"] = length
+
+        if hashes is not None:
+            self.meta["snapshot.json"]["hashes"] = hashes
```
nit: Not a huge fan of three separate lookups of "snapshot.json" key... it's not a major deal here, just pointing it out

MVrachev replied:
I imagine this would be something we would change when we fix #1333.

jku replied:
I don't think that's related: you don't need more than a single lookup even now but it is  a small detail

jku reviewed (changes requested):
Thanks. The actual API changes look fine to me
* length+hashes commit message is misleading in that this does not seem like a breaking change to me (looks like de/serialize probably works fine right now). Am I missing something?
* I prefer no TODO comments for existing issues (without a very good reason). If the issue is closed as wontfix, who fixes the TODOs?
* is there a benefit to using restructuredtext in docstrings -- do you (or are we planning to) run this through sphinx somewhere?

As a process comment I think multiple PRs instead of a single one might have increased the throughput here: now we may be stuck discussing a comment about _type when we could just merge the API changes.

MVrachev commented:
Rebased because of conflicts with develop.
Updates include:
- added additional tests for (de)serialization without hashes and length for Timestamp and Snapshot
- removed the commit with the TODO comments
- changed the length and hashes commit message

jku commented on tuf/api/metadata.py:
```diff
@@ -504,16 +504,33 @@ class Root(Signed):
             supports consistent snapshots.
         keys: A dictionary that contains a public key store used to verify
             top level roles metadata signatures::
-            {
-                '<KEYID>': <Key instance>,
-                ...
-            },
```
is this change of keyid content intentional? I think the opposite change was just done a week ago

MVrachev replied:
You are right, this is an unintentional change coming from the time when I had those changes before we added Key and Role classes.
Noticed the same for Roles.
Will fix it.
Great catch!

jku commented:
I think it looks good but I  wanted to check if the documentation change is intentional? it's not mentioned in the commit message and the opposite change was done quite recently I think
